### ISSUE TITLE
Wild variation of mAP

### ISSUE THREAD

**Issue by CdAB63** [AlexeyAB/darknet#6603]
I´m training a detector to identify firearms to be used in a low end embedded processing system attached to surveilance cameras. Basically it´s an ARM (aarch64) quad-core system without GPU acceleration and only 2GB low power DDR3 RAM. Yeah, yeah, I know... sometimes people imagine we´re magicians... But I used it for other detections using YoloV3 and it did reasonably well. Furthermore, for surveillance cameras in this particular application 1fps is OK.

Currently I have a training set of 24000 images (BGR JPEG, height=720pt, variable width). I have a dissociated test set (images from other sources) of 600 images. The training set consists of photos of firearms (2000), photos of people handling firearms (2000), photos of people handling firearms from conflict images (4000), photos of people handling firearms from streaming video (16000). Streaming video sources are surveilance cameras, police cameras, news reels, movies. The number of firearms per image ranges from 1 to more than 10 and sizes/spacial attitude/bacgrounds/etc are highly variable. The training set is complemented by 24000 negative images (images without firearms). I´m considering firearms as a single category (meaning no distinction between hand guns and long guns). Hand guns are revolvers, pistols and things like UZI (small smgs). Long guns are mostly AR systems, AKS, shotguns, bigger SMGS (H&K and the likes).

In the first attempts I used a standard YoloV3 (tiny, 3 yolo detector layers) model, then fiddled with this model to enhance detection but the increase of computational cost didn´t pay. I got an AP of about 47% which for most purposes is OK but for notification systems are highly non OK (the idea is to notify response teams and give them tactical information in order to PRESERVE lifes (minimun collateral damage)). One interesting thing is that reflexive surfaces like building glass facades, car windshields and things like that are source of false positives. Not to mention that helicopter tails are confused as AR/AK barrels and also generate false positives.

Now I´m working in the yolo v4 (again the tiny with 3 yolo layers). The procedure is to run a burnin with positve images (only images where there must be positive ID) and I let a particularly high learning rate (0.0025). As I´m working with 24000 images, I allow 52000 epochs for burn in. Then I enter the second phase where the negative images (images without gun) are introduced and learning rate is reduced to 0.001. I let the system train up to 182000 epochs. Then I introduce adversarial training and let it go to the end (about 260000 epochs).

Thing is, during the training the mAP shows the behavior ilustrated in the annex plot: it varies wildly until eventually sets to a range (in this case mAP ~51%). Is it a normal behavior? Because in other sets (not guns) it increases continuously till saturation. This attached image was obtained form a single run (no burnin, negative images from begining, no adversarial training).

One thing I´m considering is to add a deformable convolutional layer, but as darknet does not have it implemented I´m starting to write it.

![chart-20200819](https://user-images.githubusercontent.com/28064711/92151944-4bf15180-edf8-11ea-99d9-cd8cafb5b569.png)
